opm/autodiff/StandardWellsDense_impl.hpp

```diff
@@ -2539,16 +2619,268 @@ namespace Opm {
+                    const double& thp    = well_controls_iget_target(well_control, ctrl_index);
+                    const double& alq    = well_controls_iget_alq(well_control, ctrl_index);
+
+                    // Calculating the BHP value based on THP
+                    const int first_perf = wells().well_connpos[well_index]; //first perforation
+
+                    const WellType& well_type = wells().type[well_index];
+                    if (well_type == INJECTOR) {
+                        const double dp = wellhelpers::computeHydrostaticCorrection(
+                                          wells(), well_index, vfp_properties_->getInj()->getTable(vfp)->getDatumDepth(),
+                                          wellPerforationDensities()[first_perf], gravity_);
+                        const double bhp_calculated = vfp_properties_->getInj()->bhp(vfp, aqua, liquid, vapour, thp) - dp;
+                        // apply the strictest of the bhp controlls i.e. smallest bhp for injectors
+                        if (bhp_calculated < bhp) {
+                            bhp = bhp_calculated;
+                        }
+                    }
+                    else if (well_type == PRODUCER) {
+                        const double dp = wellhelpers::computeHydrostaticCorrection(
+                                          wells(), well_index, vfp_properties_->getProd()->getTable(vfp)->getDatumDepth(),
+                                          wellPerforationDensities()[first_perf], gravity_);
+                        const double bhp_calculated = vfp_properties_->getProd()->bhp(vfp, aqua, liquid, vapour, thp, alq) - dp;
+                        // apply the strictest of the bhp controlls i.e. largest bhp for producers
+                        if (bhp_calculated > bhp) {
+                            bhp = bhp_calculated;
+                        }
+                    } else {
+                       OPM_THROW(std::logic_error, "Expected PRODUCER or INJECTOR type of well");
+                    }
+                }
+            }
+
+            // there should be always some avaible bhp/thp constraints there
+            if (std::isinf(bhp) || std::isnan(bhp)) {
+                OPM_THROW(std::runtime_error, "Unvalid bhp value obtained during the potential calculation for well " << wells().name[well_index]);
+            }
+
+            converged = std::abs(old_bhp - bhp) < bhp_tolerance;
+
+            computeWellRatesWithBhp(ebosSimulator, bhp, well_index, potentials);
+
+            if (!converged) {
+                old_bhp = bhp;
+                for (int p = 0; p < np; ++p) {
+                    // TODO: improve the interpolation, will it always be valid with the way below?
+                    // TODO: finding better paramters, better iteration strategy for better convergence rate.
+                    potentials[p] = 0.001 * potentials[p] + 0.999 * old_potentials[p];
```

Comment:
Use named constant for damping factor, `potential_update_damping` and `1.0 - potential_update_damping`.